### PR TITLE
Size CurlHandler's SendTransferState based on content length

### LINKED ISSUE #7884: Determine if CurlHandler request buffer size should be optimized to start smaller and grow
From CurlHandler.EasyRequest.cs:

```csharp
                internal readonly byte[] _buffer = new byte[RequestBufferSize]; // PERF TODO: Determine if this should be optimized to start smaller and grow
```

### PR DESCRIPTION
**Pull request by stephentoub**
libcurl will never request more than 16K at a time from a request content stream, so our transfer state object allocates a 16K buffer to use the first time data needs to be transferred.  This size can be efficient in limiting the number of operations required in passing data around, but for small posts, it's an unnecessarily large size.  This commit changes the code to inspect the request content length, and if it's known to be less than 16K, that size is used instead.

cc: @ericeil, @davidsh 
Fixes #7884

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs [hunk 1]
```diff
@@ -918,7 +918,13 @@ private static ulong TransferDataFromRequestStream(IntPtr buffer, int length, Ea
                 else // sts == null
                 {
                     // Allocate a transfer state object to use for the remainder of this request.
-                    easy._sendTransferState = sts = new EasyRequest.SendTransferState();
+                    Debug.Assert(easy._requestMessage.Content != null, "Content shouldn't be null, since we already got a content request stream");
+                    long bufferSize = easy._requestMessage.Content.Headers.ContentLength.GetValueOrDefault();
+                    if (bufferSize <= 0 || bufferSize > RequestBufferSize)
+                    {
+                        bufferSize = RequestBufferSize;
```
Comment (ericeil) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Would it make sense to rename `RequestBufferSize` to something like `MaxRequestBufferSize`?

**Comment by ericeil** [comment]
LGTM, other than the one nit.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Renamed.  Thanks.

**Comment by stephentoub** [comment]
Test Innerloop OSX Debug Build and Test please
Test Innerloop OSX Release Build and Test please
Test Innerloop Ubuntu14.04 Debug Build and Test please
Test Innerloop Ubuntu14.04 Release Build and Test please

**Pull request merged**